core/modules/replica/SqlJob.h

```diff
@@ -0,0 +1,223 @@
+#define LSST_QSERV_REPLICA_SQLJOB_H
+
+// System headers
+#include <functional>
+#include <list>
+#include <map>
+#include <string>
+#include <vector>
+
+// Qserv headers
+#include "replica/Job.h"
+#include "replica/SqlRequest.h"
+#include "replica/SqlResultSet.h"
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+/**
+ * The structure SqlJobResult represents a combined result received
+ * from worker services upon a completion of the job.
+ */
+struct SqlJobResult {
+
+    /// Per-worker flag indicating if the query has succeeded at the worker
+    std::map<std::string, bool> workers;
+
+    /// Result sets for the workers
+    std::map<std::string, SqlResultSet> resultSets;
+};
+
+/**
+ * Class SqlJob represents a tool which will broadcast the same query to all
+ * worker databases of a setup. Result sets are collected in the above defined
+ * data structure.
+ */
+class SqlJob : public Job  {
+
+public:
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<SqlJob> Ptr;
+
+    /// The function type for notifications on the completion of the request
+    typedef std::function<void(Ptr)> CallbackType;
+
+    /// @return default options object for this type of a request
+    static Job::Options const& defaultOptions();
+
+    /// @return the unique name distinguishing this class from other types of jobs
+    static std::string typeName();
+
+    /**
+     * Static factory method is needed to prevent issue with the lifespan
+     * and memory management of instances created otherwise (as values or via
+     * low-level pointers).
+     *
+     * @param query
+     *   the query to be executed on all workers
+     *
+     * @param user
+     *   the name of a database account for connecting to the database service
+     *
+     * @param password
+     *   a database for connecting to the database service
+     *
+     * @param maxRows
+     *   (optional) limit for the maximum number of rows to be returned with the request.
+     *   Laving the default value of the parameter to 0 will result in not imposing any
+     *   explicit restrictions on a size of the result set. NOte that other, resource-defined
+     *   restrictions will still apply. The later includes the maximum size of the Google Protobuf
+     *   objects, the amount of available memory, etc.
+     *
+     * @param allWorkers
+     *   engage all known workers regardless of their status. If the flag
+     *   is set to 'false' then only 'ENABLED' workers which are not in
+     *   the 'READ-ONLY' state will be involved into the operation.
+     *
+     * @param controller
+     *   is needed launching requests and accessing the Configuration
+     *
+     * @param parentJobId
+     *   (optional) identifier of a parent job
+     *
+     * @param onFinish
+     *   a callback function to be called upon a completion of the job
+     *
+     * @param options
+     *   defines the job priority, etc.
```

Comment:
looks like this one is optional too. The other is not maybe?

Reply:
Yes, it is. Fixed.